pkg/agent/healthz.go

```diff
@@ -0,0 +1,65 @@
+/*
+copyright 2023 the kubernetes authors.
+
+licensed under the apache license, version 2.0 (the "license");
+you may not use this file except in compliance with the license.
+you may obtain a copy of the license at
+
+    http://www.apache.org/licenses/license-2.0
+
+unless required by applicable law or agreed to in writing, software
+distributed under the license is distributed on an "as is" basis,
+without warranties or conditions of any kind, either express or implied.
+see the license for the specific language governing permissions and
+limitations under the license.
+*/
+
+package agent
+
+import (
+	"fmt"
+	"net/http"
+)
+
+type HealthChecker interface {
+	Name() string
+	Check(req *http.Request) error
+}
+
+// ping implements the simplest possible healthz checker.
+type ping struct{}
+
+var PingHealthz HealthChecker = &ping{}
+
+func (p *ping) Name() string {
+	return "ping"
+}
+
+// PingHealthz is a health check that returns true.
+func (p *ping) Check(_ *http.Request) error {
+	return nil
+}
+
+type serverConnected struct {
+	rm ReadinessManager
+}
+
+var ServerConnectedHealthz HealthChecker = &serverConnected{}
+
+func NewServerConnectedHealthz(cs ReadinessManager) HealthChecker {
+	return &serverConnected{
+		rm: cs,
+	}
+}
+
+func (s *serverConnected) Name() string {
+	return "server-connected"
+}
+
+func (s *serverConnected) Check(_ *http.Request) error {
+	if s.rm.Ready() {
+		return nil
+	}
+
+	return fmt.Errorf("no servers connected")
```

Comment:
Healthz/liveness are used to determine if the kubelet is going to restart the pod. I don;t think we want to defaulting the agent to being restarted by kubelet because it hasn't been able to connect to konnectivity server. If we want to add that as a configurable behavior (turned on by flag) I could see the value but I don't think it should be the default behavior.

Reply:
We are not changing the behavior of the healthz/liveness probe. This only changes the behavior of readiness probe.

This too is not behind a flag, so if the flag is enabled, we have two readiness checks (ping, atleast one server connected) else only a ping check (which is quite similar to what we had before)

/cc @jkh52

Reply:
> We are not changing the behavior of the healthz/liveness probe. This only changes the behavior of readiness probe.

Looking at the change to server.go I agree. But can you make small adjustments to make it more obvious to future readers? For example, only introduce readiness.go (not healthz.go), and avoid code naming like "HealthZ" (which some readers may think implies liveness health).

Reply:
@jkh52 Sorry I missed this.

So two points if I understand correctly:
1. keep one file `readiness.go` , merge `healthz.go` into readiness.go
2. PingHealthz -> Ping / PingCheck/ PingHealth ? and similar places replacing Healtz.

I'm okay with just  Ping / ServerConnected or PingCheck  / ServerConnectedCheck for both the readiness checks.

Reply:
Sounds reasonable.